### PR TITLE
[ISSUE #3430]Refactor eventmesh-metrics-prometheus module

### LINKED ISSUE #3430: [Enhancement] Refactor eventmesh-metrics-prometheus module
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/eventmesh/issues?q=is%3Aissue) and found no similar issues.

### Enhancement Request

 Refactor eventmesh-metrics-prometheus module. 
 

### Describe the solution you'd like

1. **The current project is using a lower version of OpenTelemetry components, version 1.3.0, while the latest version is 1.24.0. The related versions need to be upgraded.**
2. **The version upgrade brings new specifications and features, requiring code refactoring.**
3. **The instrumentation of metrics needs to be adjusted accordingly based on the related semantic conversion specifications of OpenTelemetry. (https://opentelemetry.io/docs/reference/specification/metrics/semantic_conventions/)**
4. **Optimize and refactor the related code logic to make the code clearer.**

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by mxsm**
Fixes #3430 
### Motivation

1. **The current project is using a lower version of OpenTelemetry components, version 1.3.0, while the latest version is 1.24.0. The related versions need to be upgraded.**
2. **The version upgrade brings new specifications and features, requiring code refactoring.**
3. **The instrumentation of metrics needs to be adjusted accordingly based on the related semantic conversion specifications of OpenTelemetry. (https://opentelemetry.io/docs/reference/specification/metrics/semantic_conventions/)**
4. **Optimize and refactor the related code logic to make the code clearer.**

### Modifications

*Refactor eventmesh-metrics-prometheus module*
image

### Documentation

- Does this pull request introduce a new feature? (yes / no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
- If a feature is not applicable for documentation, explain why?
- If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by mxsm** [comment]
I will try fix the License Check not pass and resubmit later

**Comment by mxsm** [comment]
I will refactor this pr code and resubmit

**Comment by mxsm** [comment]
@xwm1992  PTAL~

**Comment by Pil0tXia** [comment]
@pandaapo @horoc
Refactoring the `org.apache.eventmesh.runtime.admin` package into the `eventmesh-admin` module is one of the pending tasks on our to-do list. It is advisable to merge this PR before the refactoring takes place, in order to avoid potential conflicts. PTAL at your convenience. 😉

**Comment by pandaapo** [comment]
@Pil0tXia I'm sorry about that I'm not familiar with this module for now, and in fact I do not have permission to perform an effective approving or decide merging a PR.

**Comment by mxsm** [comment]
> @pandaapo @horoc Refactoring the `org.apache.eventmesh.runtime.admin` package into the `eventmesh-admin` module is one of the pending tasks on our to-do list. It is advisable to merge this PR before the refactoring takes place, in order to avoid potential conflicts. PTAL at your convenience. 😉

@Pil0tXia There will be a biweekly meeting on Thursday where I can bring up this issue. Since it involves some refactoring, I will explain the problem during the meeting. Meanwhile, I will provide the corresponding documentation to the community as soon as possible. Additionally, everyone can help by reviewing the code. Let's aim to have the community merge this pull request as soon as possible.

**Comment by mxsm** [comment]
> @Pil0tXia I'm sorry about that I'm not familiar with this module for now, and in fact I do not have permission to perform an effective approving or decide merging a PR.
> 
> If you are worried about potential conflicts during the refactoring of "eventmesh admin", you can try asking for advice from maintainers. They may advise you to refactor first.

@pandaapo The refactoring of the admin on his side is not a major issue, and my pull request may be merged before his. He can simply rebase his branch onto the master branch when the time comes, but this process may involve resolving conflicts. The overall changes related to observability are quite significant. The main upgrades are due to version changes, resulting in differences from the previous approach. There is also a re-design of the top-level interface, among other things.

**Comment by pandaapo** [comment]
> He can simply rebase his branch onto the master branch when the time comes, but this process may involve resolving conflicts.

Do you mean that the development of refactoring "admin" can be concurrent with the rest works of this PR?

**Comment by Pil0tXia** [comment]
Concurrent work is technically feasible, while refactoring after this PR is merged will have fewer conflicts.

I will start refactoring maybe 1 month later. Not so urgent.

**Comment by mxsm** [comment]
[chinese document](https://blog.ljbmxsm.com/docs/event-mesh/core/quick-start/metric) about metrics

**Comment by Pil0tXia** [comment]
@mxsm Hi~eventmesh master often throws this exception during normal operation, a feedback to you~

image

**Comment by mxsm** [comment]
> @mxsm Hi~ EventMesh master often throws this exception during normal operation, a feedback to you~
> 
> image

Thanks for your feedback, I will try to find it and fix

**Comment by Pil0tXia** [comment]
Any progress we can make here?

**Comment by mxsm** [comment]
> Any progress we can make here?

@Pil0tXia 
I will resolve the conflicts. Can you help review the code? I've provided explanations for the code previously.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/EventMeshHttpBootstrap.java [hunk 1]
```diff
@@ -43,6 +43,7 @@ public void init() throws Exception {
         // server init
         if (eventMeshHttpConfiguration != null) {
             eventMeshHttpServer = new EventMeshHTTPServer(eventMeshServer, eventMeshHttpConfiguration);
+            eventMeshHttpServer.init();
         }
```
Comment (Pil0tXia) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/EventMeshHttpBootstrap.java:47]: Is it necessary to init eventMeshHttpServer here?

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/EventMeshServer.java [hunk 2]
```diff
@@ -96,6 +104,12 @@ public EventMeshServer() {
         if (BOOTSTRAP_LIST.isEmpty()) {
             BOOTSTRAP_LIST.add(new EventMeshTcpBootstrap(this));
         }
+        List<String> metricsPluginTypes = configuration.getEventMeshMetricsPluginType();
+        if (CollectionUtils.isNotEmpty(metricsPluginTypes)) {
+            List<MetricsRegistry> metricsRegistries = metricsPluginTypes.stream().map(metric -> MetricsPluginFactory.getMetricsRegistry(metric))
+                .collect(Collectors.toList());
+            eventMeshMetricsManager = new EventMeshMetricsManager(metricsRegistries);
+        }
```
Comment (Pil0tXia) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/EventMeshServer.java:112]: I think that maybe it is not a good design to place a list of metrics plugin types here. Is there a similar way to init metrics like other plugins (L79~L83)?

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/http/processor/HeartBeatProcessor.java [hunk 3]
```diff
@@ -147,50 +156,54 @@ public void processRequest(final ChannelHandlerContext ctx, final AsyncContext<H
...
-                    if (log.isDebugEnabled()) {
-                        log.debug("{}", httpCommand);
+
+            final CompleteHandler<HttpCommand> handler = new CompleteHandler<HttpCommand>() {
+                @Override
+                public void onResponse(final HttpCommand httpCommand) {
+                    try {
+                        if (log.isDebugEnabled()) {
+                            log.debug("{}", httpCommand);
+                        }
```
Comment (Pil0tXia) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/http/processor/HeartBeatProcessor.java:184]: `isDebugEnabled` condition is redundant.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/http/processor/HeartBeatProcessor.java [hunk 4]
```diff
@@ -68,37 +67,44 @@ public HeartBeatProcessor(final EventMeshHTTPServer eventMeshHTTPServer) {
...
-                    EventMeshRetCode.EVENTMESH_PROTOCOL_HEADER_ERR, null, HeartbeatResponseBody.class);
+        if (StringUtils.isBlank(heartbeatRequestHeader.getIdc())
+            || StringUtils.isBlank(heartbeatRequestHeader.getPid())
+            || !StringUtils.isNumeric(heartbeatRequestHeader.getPid())
+            || StringUtils.isBlank(heartbeatRequestHeader.getSys())) {
+            responseEventMeshCommand = asyncContext.getRequest().createHttpCommandResponse(
+                heartbeatResponseHeader,
+                HeartbeatResponseBody.buildBody(EventMeshRetCode.EVENTMESH_PROTOCOL_HEADER_ERR.getRetCode(),
+                    EventMeshRetCode.EVENTMESH_PROTOCOL_HEADER_ERR.getErrMsg()));
+            asyncContext.onComplete(responseEventMeshCommand);
```
Comment (Pil0tXia) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/http/processor/HeartBeatProcessor.java:94]: It seems that the main change here is the invocation of `createHttpCommandResponse`. What are the advantages of the validation logic here compared to the original? Bringing a more comprehensive validation logic?

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/metrics/MetricInstrumentUnit.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package org.apache.eventmesh.runtime.metrics;
+
+
+public abstract class MetricInstrumentUnit {
+
+    private MetricInstrumentUnit() {
+
+    }
```
Comment (Pil0tXia) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/metrics/MetricInstrumentUnit.java:25]: We can put this constant class into `constant` package together with `MonitorMetricConstants`.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/metrics/tcp/TcpMetricsCalculator.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+class TcpMetricsCalculator {
+
+    private static final int period = 30 * 1000;
```
Comment (Pil0tXia) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/metrics/tcp/TcpMetricsCalculator.java:43]: Why TCP needs a "Calculator" comparing to HTTP and gRPC?

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/metrics/MetricsManager.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.List;
+
+/**
+ * metric manager interface
+ */
+
+/**
+ * MetricsManager is an interface for managing metrics.
+ */
+public interface MetricsManager {
```
Comment (Pil0tXia) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/metrics/MetricsManager.java:31]: duplicate javadoc

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/metrics/GeneralMetricsManager.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+
+import lombok.experimental.UtilityClass;
+
+/**
+ * Managing general metrics.
+ */
+@UtilityClass
+public class GeneralMetricsManager {
```
Comment (Pil0tXia) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/metrics/GeneralMetricsManager.java:37]: The naming of `GeneralMetricsManager` is a little confusing comparing with `MetricsManager`. How about putting them in different packages and rename it to a more suitable name?

**Review by Pil0tXia — COMMENTED** [review]
Overall, I didn't see any major vulnerabilities, but there are quite a few code additions related to metrics in the EventMeshServer, which doesn't seem very maintainable. 

Additionally, the `org.apache.eventmesh.runtime.metrics` and `org.apache.eventmesh.metrics.api.model` packages could be made clearer.

**Comment by mxsm** [comment]
@Pil0tXia This PR has been open for too long, and many code changes make resolving conflicts during refactoring challenging. I will submit a new PR based on the latest code. The overall design will still be based on the current PR, and I will address the issues you mentioned above one by one.

**Comment by mxsm** [comment]
This pr will be closed, the new is #4840

**Pull request closed without merging**